File: app/api/gtt.py
"""
GTT Order API Router
====================

FastAPI endpoints for GTT (Good Till Triggered) orders.
Place, Modify, Cancel, and Get GTT orders.

Author: Antony HFT System
"""

from fastapi import APIRouter, HTTPException
from app.models.gtt import GTTPlaceRequest, GTTModifyRequest, GTTCancelRequest
from app.services.gtt_service import GTTService

router = APIRouter(prefix="/gtt", tags=["GTT Orders"])


@router.post("/place")
async def place_gtt_order(request: GTTPlaceRequest):
    """
    Place GTT Order
    
    Creates a GTT order with Entry + Target + Stop-Loss.
    
    Example Request:
    ```json
    {
        "type": "MULTIPLE",
        "quantity": 25,
        "product": "I",
        "rules": [
            {"strategy": "ENTRY", "trigger_type": "ABOVE", "trigger_price": 200},
            {"strategy": "TARGET", "trigger_type": "IMMEDIATE", "trigger_price": 220},
            {"strategy": "STOPLOSS", "trigger_type": "IMMEDIATE", "trigger_price": 190}
        ],
        "instrument_token": "NSE_FO|61755",
        "transaction_type": "BUY"
    }
    ```
    
    With Trailing SL:
    ```json
    {
        "strategy": "STOPLOSS",
        "trigger_type": "IMMEDIATE", 
        "trigger_price": 190,
        "trailing_gap": 5
    }
    ```
    """
    try:
        result = await GTTService.place_order(request)
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.put("/modify")
async def modify_gtt_order(request: GTTModifyRequest):
    """
    Modify GTT Order
    
    Update quantity, trigger prices, or add trailing SL.
    
    Example Request:
    ```json
    {
        "gtt_order_id": "GTT-C25280200137522",
        "type": "MULTIPLE",
        "quantity": 50,
        "rules": [
            {"strategy": "ENTRY", "trigger_type": "ABOVE", "trigger_price": 210},
            {"strategy": "TARGET", "trigger_type": "IMMEDIATE", "trigger_price": 230},
            {"strategy": "STOPLOSS", "trigger_type": "IMMEDIATE", "trigger_price": 195, "trailing_gap": 5}
        ]
    }
    ```
    """
    try:
        result = await GTTService.modify_order(request)
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.delete("/cancel")
async def cancel_gtt_order(request: GTTCancelRequest):
    """
    Cancel GTT Order
    
    Example Request:
    ```json
    {
        "gtt_order_id": "GTT-C25280200137522"
    }
    ```
    """
    try:
        result = await GTTService.cancel_order(request.gtt_order_id)
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/{gtt_order_id}")
async def get_gtt_order(gtt_order_id: str):
    """
    Get GTT Order Details
    
    Returns order status, rules, and execution details.
    """
    try:
        result = await GTTService.get_order(gtt_order_id)
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/")
async def get_all_gtt_orders():
    """
    Get All GTT Orders
    
    Returns list of all GTT orders for the authenticated user.
    """
    try:
        result = await GTTService.get_all_orders()
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

File: frontend/src/lib/stores/gtt.ts
// GTT Store - Good Till Triggered orders
const API_BASE = 'http://localhost:8000';

export type GTTType = 'SINGLE' | 'MULTIPLE';
export type TriggerType = 'ABOVE' | 'BELOW' | 'IMMEDIATE';
export type Strategy = 'ENTRY' | 'TARGET' | 'STOPLOSS';

export interface GTTRule {
    strategy: Strategy;
    trigger_type: TriggerType;
    trigger_price: number;
    trailing_gap?: number;
}

export interface GTTPlaceRequest {
    type: GTTType;
    quantity: number;
    product: 'I' | 'D';
    instrument_token: string;
    transaction_type: 'BUY' | 'SELL';
    rules: GTTRule[];
}

export interface GTTModifyRequest {
    gtt_order_id: string;
    type?: GTTType;
    quantity?: number;
    rules?: GTTRule[];
}

// Place GTT Order
export async function placeGTT(request: GTTPlaceRequest) {
    try {
        const res = await fetch(`${API_BASE}/api/v1/gtt/place`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify(request)
        });
        return await res.json();
    } catch (err) {
        console.error('placeGTT failed:', err);
        return { error: String(err) };
    }
}

// Modify GTT Order
export async function modifyGTT(request: GTTModifyRequest) {
    try {
        const res = await fetch(`${API_BASE}/api/v1/gtt/modify`, {
            method: 'PUT',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify(request)
        });
        return await res.json();
    } catch (err) {
        console.error('modifyGTT failed:', err);
        return { error: String(err) };
    }
}

// Cancel GTT Order
export async function cancelGTT(gttOrderId: string) {
    try {
        const res = await fetch(`${API_BASE}/api/v1/gtt/cancel`, {
            method: 'DELETE',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ gtt_order_id: gttOrderId })
        });
        return await res.json();
    } catch (err) {
        console.error('cancelGTT failed:', err);
        return { error: String(err) };
    }
}

// Get GTT Order Details
export async function getGTT(gttOrderId: string) {
    try {
        const res = await fetch(`${API_BASE}/api/v1/gtt/${gttOrderId}`);
        return await res.json();
    } catch (err) {
        console.error('getGTT failed:', err);
        return null;
    }
}

// Get All GTT Orders
export async function getAllGTT() {
    try {
        const res = await fetch(`${API_BASE}/api/v1/gtt/`);
        return await res.json();
    } catch (err) {
        console.error('getAllGTT failed:', err);
        return [];
    }
}
